x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java

```diff
@@ -93,34 +105,126 @@ protected void readInternal(byte[] b, int offset, int length) throws IOException
+        final long currentSequentialReadSize = sequentialReadSize;
+        if (currentSequentialReadSize != NO_SEQUENTIAL_READ_OPTIMIZATION) {
+            final StreamForSequentialReads streamForSequentialReads = streamForSequentialReadsRef.get();
+            if (streamForSequentialReads == null) {
+                // start a new sequential read
+                if (tryReadAndKeepStreamOpen(part, pos, b, offset, length, currentSequentialReadSize)) {
+                    return;
+                }
+            } else if (streamForSequentialReads.part == part && streamForSequentialReads.pos == pos) {
+                // continuing a sequential read that we started previously
+                assert streamForSequentialReads.isFullyRead() == false;
+                int read = streamForSequentialReads.inputStream.read(b, offset, length);
+                assert read <= length : read + " vs " + length;
+                streamForSequentialReads.pos += read;
+                position += read;
+                pos += read;
+                offset += read;
+                length -= read;
+
+                if (streamForSequentialReads.isFullyRead()) {
+                    if (streamForSequentialReadsRef.compareAndSet(streamForSequentialReads, null) != false) {
+                        streamForSequentialReads.close();
+                    } else {
+                        // something happened concurrently, defensively stop optimizing
+                        sequentialReadSize = NO_SEQUENTIAL_READ_OPTIMIZATION;
+                    }
+
+                    if (length == 0) {
+                        // the current stream contained precisely enough data for this read, so we're good.
+                        return;
+                    } else {
+                        // the current stream didn't contain enough data for this read, so we must read more
+                        if (sequentialReadSize != NO_SEQUENTIAL_READ_OPTIMIZATION
+                            && tryReadAndKeepStreamOpen(part, pos, b, offset, length, currentSequentialReadSize)) {
+                            return;
+                        }
+                    }
+                } else {
+                    // the current stream contained enough data for this read and more besides, so we leave it alone.
+                    assert length == 0 : length + " remaining";
+                    return;
+                }
+            } else {
+                // not a sequential read, so stop optimizing for this usage pattern and fall through to the unoptimized behaviour
+                assert streamForSequentialReads.isFullyRead() == false;
+                sequentialReadSize = NO_SEQUENTIAL_READ_OPTIMIZATION;
+                IOUtils.close(streamForSequentialReadsRef.getAndSet(null));
+            }
+        }
+
+        // read part of a blob directly; the code above falls through to this case where there is no optimization possible
```

Comment:
maybe put everything above this into a `readOptimized()` method that returns a boolean (denoting whether it read or not). This will allow having so many explicit returns in the above code (and the deliberate fall-through logic).

Reply:
Sounds good.

Reply:
See 1d692930391c32e0ca4f2bf0fd6d0e6fbbce7ec6 and 184c9b13764642baa3a6daf9ccabce54e4ef24d5.